### PR TITLE
[aotd] Unwrap unseen AsyncCollectiveTensor tangents

### PR DESCRIPTION
**Pull request by IvanKobzarev**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #138731

### CODE REVIEW

File: torch/_functorch/_aot_autograd/runtime_wrappers.py
Comment (bdhirsh) [inline torch/_functorch/_aot_autograd/runtime_wrappers.py:1446]: So right now we have two separate functions for doing "coercion" of tangents:

(1) coerce non-contiguous tangents to be contiguous

(2) coerce tangents with incorrect subclass metadata to have the right subclass metadata

It seems strange to add the logic for handling `AsyncCollectiveTensor` to the first one. So I would rather put it in the second (which might require tweaking the contiguous() coercion to run later), or in the worst case, merging the two pieces of coercion code.

File: torch/_functorch/_aot_autograd/runtime_wrappers.py
Comment (bdhirsh) [inline torch/_functorch/_aot_autograd/runtime_wrappers.py:1452]: hmm, I don't think it's possible to end up with an `AsyncCollectiveTensor(AsyncCollectiveTensor(...)` (that would probably a bug with the subclass itself). So I'd rather just not have a while loop here. Is there a place where you saw this show up?

**Comment by Chillee** [comment]
I don't really like that we need to shove all of this stuff into AOTAutograd. Can we register this with the subclass in some manner?

File: torch/_functorch/_aot_autograd/runtime_wrappers.py
Comment (bdhirsh) [inline torch/_functorch/_aot_autograd/runtime_wrappers.py:1450]: once you do the refactor to merge all of the tangent coercion logic into one function, you should be able to use `CompiledFunction.metadata.traced_tangent_metas` to know when you need to do a `wait()` call here instead.

File: test/functorch/test_aotdispatch.py
Comment (bdhirsh) [inline test/functorch/test_aotdispatch.py:6189]: nit: it's starting to feel like we have a lot of one-off subclasses across our tests. Any reason not to just re-use `TwoTensor` here?

**Review by bdhirsh — APPROVED** [review]
Stamping to unblock Andrew, agreed offline that we should do a refactor to merge all of the tangent coercion logic into one place (it's a bit weird to be doing `AsyncCollectiveTensor` unwrapping right now in the code that is supposed to make tangents contiguous)

File: torch/_functorch/_aot_autograd/runtime_wrappers.py
Comment (IvanKobzarev) [reply torch/_functorch/_aot_autograd/runtime_wrappers.py:1450]: Yes, we should use subclass_type information

File: test/functorch/test_aotdispatch.py
Comment (IvanKobzarev) [reply test/functorch/test_aotdispatch.py:6189]: > nit: it's starting to feel like we have a lot of one-off subclasses across our tests. Any reason not to just re-use `TwoTensor` here?

I was thinking to extract it to some utils module. Will try to use TwoTensor, I think it should work, I will try.

Comment (IvanKobzarev) [reply test/functorch/test_aotdispatch.py]: TwoTensor works, removed redundant subclass in the test.

**Comment by guilhermeleobas** [comment]
This now requires PyTorch to be built with `USE_DISTRIBUTED=1`. Is this the expected behavior?

**Comment by malfet** [comment]
> @pytorchbot revert -m "introduced regressions on linux-focal-cuda12.1-py3.10-gcc9-bazel-test" -c nosignal

It should have been ignored signal rather than no signal

> @pytorchbot merge -f "skip unrelated bazel build failures"

@IvanKobzarev why did you think it was unrelated to your change?

File: torch/_functorch/_aot_autograd/runtime_wrappers.py
Comment (bdhirsh) [inline torch/_functorch/_aot_autograd/runtime_wrappers.py:30]: oh @IvanKobzarev global import of `torch.distributed` won't work because it assumes that distributed has been built.

If we don't want to worry about extra runtime overhead of making the import local / happen at runtime, you could do:
    if torch.distributed.is_available():
        async_collective_type = type(torch.distributed._functional_collectives.AsyncCollectiveTensor)
    else:
        # dummy type, we know we won't have any AsyncCollectiveTensors at runtime if distributed is not available
        async_collective_type = NoneType

**Comment by IvanKobzarev** [comment]
> > @pytorchbot revert -m "introduced regressions on linux-focal-cuda12.1-py3.10-gcc9-bazel-test" -c nosignal
> 
> It should have been ignored signal rather than no signal
> 
> > @pytorchbot merge -f "skip unrelated bazel build failures"
> 
> @IvanKobzarev why did you think it was unrelated to your change?
@malfet ,

Sorry, did not find any log messages connected to the change and based that other variants passed and having no C++ changes wrongly decided that it is irrelevant.

**Comment by bdhirsh** [comment]
@Chillee yes we can do this (and probably should, I agree that hardcoding `AsyncCollectiveTensor` into aotd doesn't feel great).

The main tool we have to handle this from within the subclass today is a method you can define on your subclass:
```
def __coerce_same_metadata_as_tangent__(self, expected_metadata: Any):
    ...
```

it's called in AOTDispatcher here: https://github.com/pytorch/pytorch/blob/main/torch/_functorch/_aot_autograd/runtime_wrappers.py#L1475.

The problem is that this coercion function today assumes that we are not coercing from one tensor **type** to another: it just states that given subclass A with some metadata, it will return the same subclass, but coerced to have the metadata you wanted (e.g. changing a DTensor from Shard to Replicate).

In this case though, we want to coerce an `AsyncCollectiveTensor` to a plain tensor. We'll need to agree on the right API for this though.

The easiest API I can think of is to tweak the signature of that method to:
```
def __coerce_same_metadata_as_tangent__(self, expected_metadata: Any, expected_type: Type):
    ...
```

Where we also pass in the expected tensor type. Arbitrary conversions between subclass types don't really make much sense (say we expected a DTensor tangent, but actually got an NJT tangent). But in this case, the expected type is just `torch.Tensor`, and `AsyncCollectiveTensor` can write a trivial implementation that just invokes `wait`.

What do you both think?

**Comment by IvanKobzarev** [comment]
> __coerce_same_metadata_as_tangent__

Yes,  I think ability for subclasses to define how to cast them to different type is a good idea.

Similar to C++ using implicit type convertions for the call.

**Comment by IvanKobzarev** [comment]
@bdhirsh @Chillee 
Implemented Brian's proposal here: https://github.com/pytorch/pytorch/pull/139095

**Comment by IvanKobzarev** [comment]
Abandoning in favor of https://github.com/pytorch/pytorch/pull/139095

**Pull request closed without merging**
